File: 0216-combination-sum-iii/0216-combination-sum-iii.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function (k, n) {
    const dfs = function (remSum, start, path) {
        if(path.length === k && remSum === 0) {
            result.push([...path]);
            return;
        }

        for (let i = start; i <= 9; i++) {
            path.push(i);
            dfs(remSum - i, i + 1, path);
            path.pop();
        }
    }

    const result = [];
    dfs(n, 1, []);

    return result;
};

File: 0216-combination-sum-iii/0216-combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        def dfs(remSum, start, path):
            if len(path) == k and remSum == 0:
                result.append(path[:])
                return

            if remSum < 0:
                return

            for i in range(start, 10):
                path.append(i)
                dfs(remSum - i, i + 1, path)
                path.pop()

        result = []
        dfs(n, 1, [])

        return result
